docs/terminals.md

```diff
@@ -0,0 +1,141 @@
+
+## Which Mode To Use
+`runc` itself runs in two modes:
+
+* foreground
+* detached
+
+You can use either terminal mode with either runc mode. However, there are considerations that may indicate preference for one mode over another.
+
+### Foreground
+When running in foreground, `runc` and the container you invoke are most like running a regular foreground process. This is the scenario most likely to be useful for running in pass-through terminal mode. As the example above showed:
+
+```
+$ echo input | runc run some_container > /tmp/log.out 2>& /tmp/log.err
+```
+
+Nonetheless, you still have the option, when running in foreground runc mode, of using New Terminal mode. However, if the process is short-lived, you will not have much time to retrieve the console's fds from the socket and doing anything with it.
+
+### Detached
+When running in detached runc mode, you can run your terminal either in Pass-Through or New-Terminal mode.
+
+You run in detached mode by doing one of the following:
+
+* `runc -d`
+* `runc create` followed by `runc start`, i.e. container lifecycle management
+
+#### Pass-Through Detached
+When running runc detached with terminal Pass-Through, the stdin/stdout/stderr of the calling process will be passed to `runc` and from there to the container's process. However, since `runc` detaches, there are several side effects to take into account:
+
+* The stdout/stderr of the process can become intermixed with the stdout/stderr of the calling process.
+* Any input into the calling process can cause indeterminate issues with which process actually receives the input.
+
+For example:
+
+```
+$ runc create foo              # container created, inheriting stdio from this terminal
+$ runc start foo               # container started, inheriting stdio from this terminal
+$ curl https://containerd.io   # get contents of this site to the terminal
+
+<!DOCTYPE html>
+I am container output
+<html lang="en">
+  <head>
+	  ASashcned more container output
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>containerd</title>
+```
+
+Two problems occurred above:
+
+1. The output from the `curl` command and the output of the container `foo` became intermingled.
+2. Our typing `curl https://containerd.io` actually was passed to both our shell and the container `foo`. The results can be unpredictable. Our example here was simulated and might not even work.
+
+Further, a third, possibly temporary problem exists. Pass-Through Detached only works correctly when stdout/stderr are either an actual terminal or a file. If you use _anything_ else, e.g. a pipe, it will hang, for reasons unknown. There is an [open issue](https://github.com/opencontainers/runc/issues/1721) on this.
+
+For this reason, we **strongly recommend** using New Terminal mode when running `runc` detached.
```

Comment:
... unless you set up your own stdio outside the container.

Reply:
Good point. Fixing.